sensor_msgs/include/sensor_msgs/point_cloud_msg_wrapper.hpp

```diff
@@ -0,0 +1,496 @@
+///             in such a way that the underlying point cloud message is not deleted before the
+///             wrapper.
+///
+///             For convenience, there are two typedefs: PointCloud2View and PointCloud2Modifier.
+///             - PointCloudView<PointT>{msg} wraps a constant point cloud and allows read-only
+///               access. Modification through this view is impossible.
+///             - PointCloud2Modifier<PointT>{msg} wraps a mutable message and allows read-write
+///               access to the underlying data.
+///             - PointCloud2Modifier<PointT>{msg, new_frame_id} initializes an empty mutable
+///               message. This constructor is to be used solely to initialize a new message and
+///               will throw if the point cloud message is already initialized (i.e., has non-zero
+///               number of fields).
+///
+/// @tparam     PointT           Type of point to use for message reading/writing.
+/// @tparam     CloudMsgT        Type of point cloud message.
+/// @tparam     FieldGenerators  A tuple of all field generators that allow generating Field structs
+///                              from members of a PointT struct. See
+///                              LIDAR_UTILS__DEFINE_FIELD_GENERATOR_FOR_MEMBER for more details.
+///                              The class provides a sane default value here, but a custom tuple
+///                              can be generated by the user by using
+///                              LIDAR_UTILS__DEFINE_FIELD_GENERATOR_FOR_MEMBER if needed.
+///
+template<typename PointT, typename CloudMsgT, typename FieldGenerators>
+class PointCloudMsgWrapper
+{
+  static_assert(
+    std::is_default_constructible<PointT>::value,
+    "\n\nThe point type must satisfy is_default_constructible trait.\n\n");
+
+  static_assert(
+    detail::is_specialization<FieldGenerators, std::tuple>::value,
+    "\n\nFieldGenerators must be an std::tuple.\n\n");
+
+  static_assert(
+    std::is_same<std::decay_t<CloudMsgT>, sensor_msgs::msg::PointCloud2>::value,
+    "\n\nThis class is designed to work with PointCloud2 messages only for now.\n\n");
+
+  using FieldNameT = sensor_msgs::msg::PointField::_name_type;
+
+public:
+  /// Needed for std::back_inserter.
+  using value_type = PointT;
+
+  ///
+  /// @brief      A const iterator over the underlying point cloud message.
+  ///
+  class ConstIterator;
+  ///
+  /// @brief      A modifying iterator over the underlying point cloud message. Only relevant if the
+  ///             point cloud type is not const.
+  ///
+  class Iterator;
+
+  ///
+  /// @brief      Constructor that wraps a point cloud message. Checks if the message fields
+  ///             correspond to the ones generated from the PointT type and throws in case of a
+  ///             mismatch.
+  ///
+  /// @param      cloud_ref  A reference to the wrapped cloud.
+  ///
+  explicit PointCloudMsgWrapper(CloudMsgT & cloud_ref)
```

Comment:
Just to start a discussion. Should we allow an unsafe version of this wrapper? We could wrap it into a separate class like `IKnowWhatImDoing` and allow the user to manually specify the fields for those cases when this class is otherwise not flexible enough. The plus point - more power to the **power**-user (normal power user would use a custom struct, thus the bold text). The minus point, I can't imagine when this would be needed. We should be able to create a point struct/class with correct alignment/offsets by clever using of `alignas` etc. Thoughts?